### DISCUSSION TITLE
2010/08/18/%E8%AE%BE%E8%AE%A1%EF%BC%9A%E5%90%8C%E4%B8%80%E5%8A%9F%E8%83%BD%E5%AF%B9%E5%BA%94%E5%A4%9A%E7%A7%8D%E4%B8%8D%E5%90%8C%E7%95%8C%E9%9D%A2%E7%9A%84%E5%BA%94%E7%94%A8%E7%A8%8B%E5%BA%8F

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/blog#294] [Announcements]
# 2010/08/18/%E8%AE%BE%E8%AE%A1%EF%BC%9A%E5%90%8C%E4%B8%80%E5%8A%9F%E8%83%BD%E5%AF%B9%E5%BA%94%E5%A4%9A%E7%A7%8D%E4%B8%8D%E5%90%8C%E7%95%8C%E9%9D%A2%E7%9A%84%E5%BA%94%E7%94%A8%E7%A8%8B%E5%BA%8F

虽然我以前研究过动态注册事件这一功能（3.7.6），但是还没有想到，这个功能有什么实际用途。直到前几天，一个客户向我介绍了他们工作中遇到的困难，我才突然想到，他们的问题恰好可以使用动态注册事件来解决。

https://blog.qizhen.xyz/2010/08/18/%E8%AE%BE%E8%AE%A1%EF%BC%9A%E5%90%8C%E4%B8%80%E5%8A%9F%E8%83%BD%E5%AF%B9%E5%BA%94%E5%A4%9A%E7%A7%8D%E4%B8%8D%E5%90%8C%E7%95%8C%E9%9D%A2%E7%9A%84%E5%BA%94%E7%94%A8%E7%A8%8B%E5%BA%8F

**Comment by ruanqizhen** [comment]
之前博客下的留言：

ruanqizhen
2011年03月29日 10:55 编辑
我之前有个这相发，觉得是可行的，也没有真正给它写代码。既然你提出来需要一个范例，我今天就抽空做了一个简单的例子。我会把这个例子的介绍就添加在原文的下面。

回复
zyh7148
2011年03月29日 13:29 编辑
阮工：
你好！你给的例程是通过在界面VI中调用代码VI实现的，如果界面上的控件比较多的话，那么传给代码VI的输入参数就比较多，有没有方法不通过直接调用代码VI，而是通过控件名来传值，比如通过VI的Set Value或Get Value方法等，或有更好的方法。谢谢！

回复
ruanqizhen
2011年03月30日 08:47 编辑
你可以把界面VI的引用传给Main.vi，然后再Main.vi中通过控件名就可以得到对应控件的引用，之后就可以通过来做赋值、事件响应等的操作了。
你可以参考这篇文章里关于得到控件引用的部分：https://ruanqizhen.wordpress.com/2008/07/10/%E7%95%8C%E9%9D%A2%E8%AE%BE%E8%AE%A1%E6%8A%80%E5%B7%A7-1-%E5%88%A9%E7%94%A8-labview-%E8%87%AA%E5%B8%A6%E6%8E%A7%E4%BB%B6/
或者你有我的书的话，对应的内容在6.2.3和10.3.2节中。

回复
zyh7148
2011年03月30日 13:44 编辑
阮工：
你好！那如果通过这种方法，那在主VI中注册动态事件时，只能注册面板中所有Control Value Change事件，就不能像你例程中那样对每个控件注册一个动态事件了，只能在事件结构中再判断是哪个控件发生了变化，是这样吗？顺便问一下，在LabVIEW的安装目录下有哪几个目录下的VI是非常有用，但在程序面板中没有显示出来的？就像你的一个例程中Get Control这样的VI。你对LabVIEW的文件结构可能更了解。非常感谢！

回复
陈小刚
2011年03月30日 16:37 编辑
我喜欢把界面上的控件组合成一个一个的簇来存放，这些簇全都用严格类型定义。把这些簇的引用传入子vi再去做处理，跟用Get Control差不多，不过会稍微好用一点，毕竟你自己做过分门别类，不知道这个用法怎么样。

回复
ruanqizhen
2011年03月31日 09:25 编辑
我的意思是把界面VI的引用作为参数。再通过VI的引用，得到VI上面每个控件的引用。在Main.vi中得到Interface.vi上控件的引用之后，就既可以使用这个引用注册事件，也可以使用这个引用修改属性值了。

回复
zyh7148
2011年03月31日 11:18 编辑
阮工：
你好！可能我还是没有说明我的意思，我的意思是你在interface.vi中是把Main.vi作为子vi调用的，你的例程中interface.vi上是三个控件，你的Main.vi的输入参数就是三个，因为Main.vi中你注册动态事件时是对每个控件的引用分别注册动态事件的，因为对每个控件事件的处理是不一样的，如果interface.vi上的控件比较多时，Main.vi就要有很多输入参数，我的意思是能否有一种方法，当interface.vi上的控件比较多时，Main.vi的输入参数比较少，但又能实现你例程的功能。谢谢！

回复
ruanqizhen
2011年03月31日 13:32 编辑
我的意思是：在interface.vi与Main.vi之间，只有一个参数，是interface.vi的引用。
在Main.vi中通过这一个引用就可以得到interface.vi上所有的控件。
建议你看一下我发给你的链接。

回复
陈小刚
2011年03月31日 17:28 编辑
对，是这样，阮老师是从宏观的角度，我的角度比较微观一点:)。我的意思是，在Interface1和2里，如果控件数量很多，如果不用簇来规整的话，直接去Get Control，注册事件的时候就会非常多，比如例子里是2个按钮，就是注册两个value change，如果是10个的话就得下拉10个。如果把这10个按钮组成一个簇，注册的时候只需一次，事件处理的时候，可以写在一个事件里（cluster-All Element)。如果用簇把控件组合成几块，在Main.vi里注册和处理事件的时候就会相对少一点。另外，再把这些簇的引用做成type definition,如果interface上的控件有更改的话，修改起来也比较方便。

回复
zyh7148
2011年03月28日 17:11 编辑
阮工：
你好！我想要利用这个功能，能否给一个简单的例子，尤其是真正执行代码的那个VI，多谢了！从界面VI我是利用了VI的Set Value和Get Value方法，虽然我了解了一些动态注册事件，但代码VI我还是不知怎样去实现。

回复
ruanqizhen
2011年02月21日 12:29 编辑
使用属性节点的确比直接读写值效率低下。但是，界面不同于数学运算，在界面操作中，效率的瓶颈是人的反应速度，而不是计算机的云算速度。一个界面，假设有上千个控件，全部使用属性节点，可能效率最多低几个毫秒，这几个毫秒，人眼根本无法察觉。

回复
huangchen
2011年02月19日 23:33 编辑
阮大侠，这样做的话，因为只有控件的引用，读取和赋值都得通过属性节点，岂不是程序效率会大大降低？

回复
